File: AdguardExtension/ActionExtension/ActionViewController.m
/**
    This file is part of Adguard for iOS (https://github.com/AdguardTeam/AdguardForiOS).
    Copyright © Adguard Software Limited. All rights reserved.
 
    Adguard for iOS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
 
    Adguard for iOS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with Adguard for iOS.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "ActionViewController.h"
#import <MobileCoreServices/MobileCoreServices.h>

#import "ACommons/ACLang.h"
#import "ASDatabase/ASDatabase.h"
#import "AESharedResources.h"
#import "AEService.h"
#import "AESAntibanner.h"
#import "AEWhitelistDomainObject.h"
#import "ASDFilterObjects.h"
#import "AEInvertedWhitelistDomainsObject.h"
#import "Adguard-Swift.h"

#define USER_FRIENDLY_DELAY     0.5

NSString *AEActionErrorDomain = @"AEActionErrorDomain";
#define AE_ACTION_ERROR_NODEFAULTS     100
#define AE_ACTION_ERROR_NODB           200


@interface ActionViewController (){
    
    AESharedResources *_sharedResources;
    AEService *_aeService;
    SafariService *_safariService;
    ContentBlockerService *_contentBlockerService;
    AESSupport *_support;
    
    NSURL *_url;
    BOOL _injectScriptSupported;
    NSString *_host;
    NSURL *_iconUrl;
    BOOL _enabled;
    NSMutableArray *_observerObjects;
    ActionExtensionMainController __weak *_mainController;
}

//@property(strong,nonatomic) IBOutlet UIImageView *imageView;

@end

@implementation ActionViewController

/////////////////////////////////////////////////////////////////////
#pragma mark Class Methods
/////////////////////////////////////////////////////////////////////

+ (BOOL) isHostInInvertedWhitelist:(NSString *)host {
    
    @autoreleasepool {
        
        AEInvertedWhitelistDomainsObject *invertedDomainsObj = [AESharedResources new].invertedWhitelistContentBlockingObject;
        
        __block BOOL found = NO;
        
        [invertedDomainsObj.domains enumerateObjectsUsingBlock:^(NSString * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            
            if([obj caseInsensitiveCompare:host] == NSOrderedSame) {
                found = YES;
                *stop = YES;
            }
        }];
        
        return found;
    }
}

+ (AEWhitelistDomainObject *)domainObjectIfExistsFromContentBlockingWhitelistFor:(NSString *)host{
    
    @autoreleasepool {
        
        DDLogDebug(@"(ActionViewController) domainObjectIfExistsFromContentBlockingWhitelistFor:\"%@\"", host);
        NSArray *rules = [[AESharedResources new] whitelistContentBlockingRules];
        
        return [self domainObjectIfExists:host inRules:rules];
    }
}

+ (AEWhitelistDomainObject *)domainObjectIfExists:(NSString *)host inRules:(NSArray*) rules {
    
    rules = [rules
             filteredArrayUsingPredicate:
             [NSPredicate
              predicateWithFormat:@"ruleText CONTAINS[cd] %@",
              host]];
    
    if (rules.count) {
        
        AEWhitelistDomainObject *obj;
        for (ASDFilterRule *rule in rules) {
            obj = [[AEWhitelistDomainObject alloc] initWithRule:rule];
            if (obj) {
                break;
            }
        }
        
        return obj;
    }
    
    return nil;
}


/////////////////////////////////////////////////////////////////////
#pragma mark Public Methods
/////////////////////////////////////////////////////////////////////

- (void)viewDidLoad {
    [super viewDidLoad];
    
//    [self startProcessing];

    // Get the item[s] we're handling from the extension context.

    self.navigationController.navigationBar.shadowImage = [UIImage new];

    _sharedResources = [AESharedResources new];
    _safariService = [[SafariService alloc] initWithResources:_sharedResources];
    _contentBlockerService = [[ContentBlockerService alloc] initWithResources:_sharedResources safariService:_safariService];
    _aeService = [[AEService alloc] initWithContentBlocker:_contentBlockerService resources:_sharedResources];
    
    _support = [[AESSupport alloc] initWithResources:_sharedResources safariSevice:_safariService aeService:_aeService];

    self.title = LocalizationNotNeeded(AE_PRODUCT_NAME);

    __block NSString *errorMessage = ACLocalizedString(@"support_error_safari_extension", @"(Action Extension - ActionViewController) Some errors when starting.");

    NSExtensionItem *item = self.extensionContext.inputItems.firstObject;
    NSItemProvider *itemProvider = item.attachments.firstObject;

    ASSIGN_WEAK(self);

    if ([itemProvider hasItemConformingToTypeIdentifier:(NSString *)kUTTypePropertyList]) {
        [itemProvider loadItemForTypeIdentifier:(NSString *)kUTTypePropertyList options:nil completionHandler:^(NSDictionary *results, NSError *error) {

            ASSIGN_STRONG(self);

            NSDictionary *theDict = results[NSExtensionJavaScriptPreprocessingResultsKey];
            NSString *urlString = theDict[@"urlString"];
            if (urlString) {
                USE_STRONG(self)->_url = [NSURL URLWithString:urlString];
            }
            USE_STRONG(self)->_host = [USE_STRONG(self)->_url hostWithPort];
            //            _host = url.host;x

            USE_STRONG(self)->_injectScriptSupported = [theDict[@"injectScriptSupported"] boolValue];

            if (error) {

                DDLogError(@"(ActionViewController) Error of obtaining page url from Safari:\n%@", [error localizedDescription]);
            }
            else if ([NSString isNullOrEmpty:_host]) {

                DDLogError(@"(ActionViewController) Error of obtaining page url from Safari: url is empty.");
                errorMessage = ACLocalizedString(@"hostname_obtaining_error", @"(Action Extension - ActionViewController) Can't obtain hostname when starting.");
            }
            else {

                NSError *error = [USE_STRONG(self) prepareDataModel];
                if (error) {

                    if (error.code == AE_ACTION_ERROR_NODB) {
                        errorMessage = error.localizedDescription;
                    }
                }
                else {
                    
                    [USE_STRONG(self)->_aeService onReady:^{

                        // Add observers for application notifications
                        USE_STRONG(self)->_observerObjects = [NSMutableArray arrayWithCapacity:2];

                        id observerObject = [[NSNotificationCenter defaultCenter] addObserverForName:UIApplicationDidEnterBackgroundNotification object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {

                            [AESharedResources synchronizeSharedDefaults];
                        }];
                        if (observerObject) {
                            [USE_STRONG(self)->_observerObjects addObject:observerObject];
                        }
                        observerObject = [[NSNotificationCenter defaultCenter] addObserverForName:UIApplicationWillTerminateNotification object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {

                            [AESharedResources synchronizeSharedDefaults];
                        }];
                        if (observerObject) {
                            [USE_STRONG(self)->_observerObjects addObject:observerObject];
                        }
                        //--------------------------------------------

                        USE_STRONG(self)->_iconUrl = [NSURL URLWithString:[NSString stringWithFormat:@"%@://%@/favicon.ico", USE_STRONG(self)->_url.scheme, [USE_STRONG(self)->_url hostWithPort]]];

                        [USE_STRONG(self) startProcessing];

                    }];

                    return;
                }
            }
            //done on error
            [USE_STRONG(self) stopProcessingWithMessage:errorMessage];
        }];

        return;
    }

    //done on error
    [self stopProcessingWithMessage:errorMessage];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (void)dealloc{
    
    DDLogDebug(@"(ActionViewController) run dealloc.");
    
    for (id observer in _observerObjects) {
        [[NSNotificationCenter defaultCenter] removeObserver:observer];
    }
}

- (NSError *)prepareDataModel{
    
    // Init Logger
    [[ACLLogger singleton] initLogger:[AESharedResources sharedAppLogsURL]];
    
#if DEBUG
    [[ACLLogger singleton] setLogLevel:ACLLDebugLevel];
#endif
    
    // Registering standart Defaults
    NSString *appPath = [[[NSBundle mainBundle] bundlePath] stringByAppendingPathComponent:@"../../"];
    NSDictionary * defs = [NSDictionary dictionaryWithContentsOfFile:[[NSBundle bundleWithPath:appPath] pathForResource:@"defaults" ofType:@"plist"]];
    if (defs){
        
        DDLogInfo(@"(ActionViewController) default.plist loaded!");
        
        [[_sharedResources sharedDefaults] registerDefaults:defs];
    }
    else{
        
        DDLogError(@"(ActionViewController) default.plist was not loaded.");
        return [NSError errorWithDomain:AEActionErrorDomain
                                   code:AE_ACTION_ERROR_NODEFAULTS
                               userInfo:nil];
    }
    //-------------------------------
    
    // Init database
    NSURL *dbURL = [[AESharedResources sharedResuorcesURL] URLByAppendingPathComponent:AE_PRODUCTION_DB];

    [[ASDatabase singleton] initDbWithURL:dbURL upgradeDefaultDb:NO];

    if ([[ASDatabase singleton] error]) {
        
        DDLogError(@"(ActionViewController) production DB was not created before.");
        NSString *messageFormat =
        ACLocalizedString(@"action_extension_no_configuration_message_format",
                          @"(Action Extension - ActionViewController) An error which occurs if the action extension is launched before the main Adguard app.");
        return [NSError errorWithDomain:AEActionErrorDomain
                                   code:AE_ACTION_ERROR_NODB
                               userInfo:@{NSLocalizedDescriptionKey : [NSString stringWithFormat:messageFormat, AE_PRODUCT_NAME, AE_PRODUCT_NAME]
                                          }];
    }
    
    ASSIGN_WEAK(self);
    dispatch_async(dispatch_get_main_queue(), ^{
        
        ASSIGN_STRONG(self);
        
        //------------ Checking DB status -----------------------------
        ASDatabase *dbService = [ASDatabase singleton];
        if (dbService.error) {
            
            //        [self dbFailure];
        }
        else if (!dbService.ready){
            
            [dbService addObserver:USE_STRONG(self) forKeyPath:@"ready" options:NSKeyValueObservingOptionNew context:nil];
        }
        //--------------------- Start Services ---------------------------
        else
            [USE_STRONG(self)->_aeService start];
        
    });
    
    return nil;
}

/////////////////////////////////////////////////////////////////////
#pragma mark Navigations
/////////////////////////////////////////////////////////////////////

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender{
    
    if ([segue.identifier isEqualToString:@"loader"]) {
        
        _mainController = (ActionExtensionMainController *)segue.destinationViewController;
        
        _mainController.resources = _sharedResources;
        _mainController.safariService = _safariService;
        _mainController.contentBlockerService = _contentBlockerService;
        _mainController.support = _support;
        
        _mainController.domainName = _host;
        _mainController.url = _url;
        _mainController.iconUrl = _iconUrl;
        _mainController.domainEnabled = _enabled;
        _mainController.injectScriptSupported = _injectScriptSupported;
        
        _mainController.enableChangeDomainFilteringStatus = YES;
    }
}

/////////////////////////////////////////////////////////////////////
#pragma mark Observing notifications
/////////////////////////////////////////////////////////////////////

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context{
    
    // DB DELAYED READY
    ASDatabase * dbService = [ASDatabase singleton];
    if ([object isEqual:dbService]
        && [keyPath isEqualToString:@"ready"]
        && dbService.ready) {
        
        [dbService removeObserver:self forKeyPath:@"ready"];
        
        //--------------------- Start Services ---------------------------
        [_aeService start];
        
        return;
    }
    
    // Default processing
    [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
}

/////////////////////////////////////////////////////////////////////
#pragma mark Helper Methods (Private)
/////////////////////////////////////////////////////////////////////

- (void)stopProcessingWithMessage:(NSString *)message{
    
    dispatch_async(dispatch_get_main_queue(), ^{
        
        [self.loadIndicator stopAnimating];
        [self.loadIndicator setHidden:YES];
        if (message) {
            self.messageLabel.text = message;
        }
        [self.messageLabel setHidden:NO];
    });
}

- (void)startProcessing{
    
    dispatch_async(dispatch_get_main_queue(), ^{
        
        [self.loadIndicator startAnimating];
        [self.loadIndicator setHidden:NO];
        self.messageLabel.text = @"";
        [self.messageLabel setHidden:YES];
    });
    
    BOOL inverted = [_sharedResources.sharedDefaults boolForKey:AEDefaultsInvertedWhitelist];
    
    if(inverted) {
        
        _enabled = [ActionViewController isHostInInvertedWhitelist:_host];
    }
    else {
       
        _enabled = [ActionViewController domainObjectIfExistsFromContentBlockingWhitelistFor:_host] == nil;
    }
    
    ASSIGN_WEAK(self);
    dispatch_async(dispatch_get_main_queue(), ^{
        ASSIGN_STRONG(self);
        [USE_STRONG(self).actionButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    });
    
}

@end

File: AdguardExtension/ActionExtension/Swift UI/ActionViewController.swift
/**
    This file is part of Adguard for iOS (https://github.com/AdguardTeam/AdguardForiOS).
    Copyright © Adguard Software Limited. All rights reserved.

    Adguard for iOS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Adguard for iOS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Adguard for iOS.  If not, see <http://www.gnu.org/licenses/>.
*/

import UIKit
import CoreServices

class ActionViewController: UIViewController {

    @IBOutlet weak var messageLabel: UILabel!
    
    @IBOutlet weak var loadIndicator: UIActivityIndicatorView!
    
    // MARK: - Variables
    
    private var url: URL? = nil
    private var iconUrl: URL? = nil
    private var host: String? = nil
    private var injectScriptSupported = false
    private var enabled = false
    private var dbObserver: Any?
    private var configuration: SimpleConfigurationSwift
    
    var systemStyleIsDark: Bool {
        if #available(iOSApplicationExtension 13.0, *) {
            switch traitCollection.userInterfaceStyle {
            case .light:
                return false
            case .dark:
                return true
            default:
                return false
            }
        } else {
            return false
        }
    }
    
    // MARK: - Constants
    
    private let AEActionErrorDomain = "AEActionErrorDomain"
    private let aeProductionDb = "adguard.db"
    private let AEActionErrorNoDefaults = 100
    private let AEActionErrorNoDb = 200
    private let userFriendlyDelay = 0.5
    private let segueId = "loader"
    
    // MARK: - Services
    
    private let sharedResources: AESharedResourcesProtocol = AESharedResources()
    private let safariService: SafariService
    private let contentBlockerService: ContentBlockerService
    private let networking = ACNNetworking()
    private let antibannerController: AntibannerControllerProtocol
    private let webReporter: ActionWebReporterProtocol
    private var theme: ThemeServiceProtocol?
    private let asDataBase = ASDatabase()
    private let productInfo: ADProductInfoProtocol
    
    private var notificationToken1: NotificationToken?
    private var notificationToken2: NotificationToken?
    
    // MARK: - View Controller LifeCycle
    
    required init?(coder: NSCoder) {
        safariService = SafariService(resources: sharedResources)
        let antibanner = AESAntibanner(networking: networking, resources: sharedResources)
        configuration = SimpleConfigurationSwift(withResources: sharedResources, systemAppearenceIsDark: true)
        
        productInfo = ADProductInfo()
        
        self.antibannerController = AntibannerController(antibanner: antibanner, resources: sharedResources, productInfo: productInfo)
        
        let safariProtectoin = SafariProtectionService(resources: sharedResources)
        
        let dnsFiltersService = DnsFiltersService(resources: sharedResources, vpnManager: nil, configuration: configuration, complexProtection: nil)
        
        let safariProtection = SafariProtectionService(resources: sharedResources)
        
        let dnsProviders = DnsProvidersService(resources: sharedResources)
        
        let networkSettings = NetworkSettingsService(resources: sharedResources)

        let vpnManager = VpnManager(resources: sharedResources, configuration: configuration, networkSettings: networkSettings, dnsProviders: dnsProviders)
        
        let nativeProviders = NativeProvidersService(dnsProvidersService: dnsProviders, networkSettingsService: networkSettings, resources: sharedResources, configuration: configuration)
        
        let complexProtection = ComplexProtectionService(resources: sharedResources, safariService: safariService, configuration: configuration, vpnManager: vpnManager, safariProtection: safariProtectoin, productInfo: productInfo, nativeProvidersService: nativeProviders)
 
        contentBlockerService = ContentBlockerService(resources: sharedResources,
                                                      safariService: safariService,
                                                      antibanner: antibanner,
                                                      safariProtection: safariProtection)
        
        webReporter = ActionWebReporter(productInfo: productInfo,
                                    antibanner: antibanner,
                                    complexProtection: complexProtection,
                                    dnsProviders: dnsProviders,
                                    configuration: configuration,
                                    dnsFilters: dnsFiltersService)
        
        super.init(coder: coder)
    }
    
    deinit {
        DDLogDebug("(ActionViewController) run deinit.")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.configuration.systemAppearenceIsDark = systemStyleIsDark
        self.theme = ThemeService(configuration)
        
        navigationController?.navigationBar.shadowImage = UIImage()
        
        title = LocalizationNotNeeded(Constants.aeProductName())
        var errorMessage = ACLocalizedString("support_error_safari_extension", nil)
        guard let item: NSExtensionItem = self.extensionContext?.inputItems.first as? NSExtensionItem else { return }
        guard let itemProvider: NSItemProvider = item.attachments?.first else { return }
        
        if !itemProvider.hasItemConformingToTypeIdentifier(String(kUTTypePropertyList)) {
            DDLogError("(ActionViewController) Error  - itemProvider does not conform to type \(String(kUTTypePropertyList))")
            stopProcessing(with: errorMessage)
            return
        }
        
        itemProvider.loadItem(forTypeIdentifier: String(kUTTypePropertyList), options: nil) {[weak self] (results, error) in
            
            var succeed = false
            
            defer {
                if !succeed {
                    self?.stopProcessing(with: errorMessage)
                }
            }
            
            guard let self = self else { return }
            
            guard let dictResult = results as? Dictionary<String, Any> else {
                DDLogError("(ActionViewController) Error - result dict incorrect. Results: \(results.debugDescription )")
                return
            }
            
            guard let theDict = dictResult[NSExtensionJavaScriptPreprocessingResultsKey] as? Dictionary<String, Any> else {
                DDLogError("(ActionViewController) Error - can not get NSExtensionJavaScriptPreprocessingResultsKey. Results: \(results.debugDescription )")
                return
            }
            
            if let urlString = theDict["urlString"] as? String {
                self.url = URL(string: urlString)
            }
            
            self.host = self.url?.hostWithPort()
            
            if let supported = theDict["injectScriptSupported"] as? Int {
                self.injectScriptSupported = supported == 0 ? false : true
            }
            
            if error != nil {
                DDLogError("(ActionViewController) Error of obtaining page url from Safari:\(error?.localizedDescription ?? "nil")" )
            } else if self.host?.isEmpty ?? true || self.host == nil {
                DDLogError("(ActionViewController) Error of obtaining page url from Safari: url is empty.")
                errorMessage = ACLocalizedString("hostname_obtaining_error", "(Action Extension - ActionViewController) Can't obtain hostname when starting.")
            } else {
                
                let error = self.prepareDataModel()
                
                if error != nil {
                    DDLogError("(ActionViewController) can not prepare data model. \(error!.localizedDescription)")
                    if error?.code == self.AEActionErrorNoDb {
                        errorMessage = error?.localizedDescription ?? ""
                    }
                } else {
                    succeed = true
                    
                    self.antibannerController.onReady { (antibanner) in
                        // Add observers for application notifications
                        self.addObservers()
                        
                        let formattedString = String(format: "%@://%@/favicon.ico", self.url?.scheme ?? "", self.url?.hostWithPort() ?? "")
                        self.iconUrl = URL(string: formattedString)
                        
                        self.startProcessing()
                    }
                }
            }
        }
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        configuration.systemAppearenceIsDark = systemStyleIsDark
        updateTheme()
    }
    
    // MARK: - Prepare for segue
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == segueId {
            guard let mainVC = segue.destination as? ActionExtensionMainController else { return }
            mainVC.resources = sharedResources
            mainVC.safariService = safariService
            mainVC.contentBlockerService = contentBlockerService
            mainVC.webReporter = webReporter
            mainVC.domainName = host
            mainVC.iconUrl = iconUrl
            mainVC.domainEnabled = enabled
            mainVC.injectScriptSupported = injectScriptSupported
            mainVC.enableChangeDomainFilteringStatus = true
            mainVC.url = url
        }
    }
    
    
    // MARK: - Methods
    
    @IBAction func closeAction(_ sender: UIButton) {
        self.extensionContext?.completeRequest(returningItems: nil, completionHandler: nil)
    }
    
    
    // MARK: - Private methods
    
    private func startProcessing(){
        DispatchQueue.main.async {[weak self] in
            guard let sSelf = self else { return }
            sSelf.loadIndicator.startAnimating()
            sSelf.messageLabel.text = ""
            sSelf.messageLabel.isHidden = true
        }
        
        let inverted = sharedResources.sharedDefaults().bool(forKey: AEDefaultsInvertedWhitelist)
        
        if inverted {
            enabled = isHostInInvertedWhitelist(host: host ?? "")
        } else {
            enabled = domainObjectIfExistsFromContentBlockingWhitelistFor(host: host ?? "") == nil
        }
        
        DispatchQueue.main.async {[weak self] in
            guard let sSelf = self else { return }
            sSelf.performSegue(withIdentifier: sSelf.segueId, sender: sSelf)
        }
    }
    
    private func stopProcessing(with message: String){
        DispatchQueue.main.async {[weak self] in
            self?.loadIndicator.stopAnimating()
            if !message.isEmpty {
                self?.messageLabel.text = message
            }
            self?.messageLabel.isHidden = false
        }
    }
    
    private func addObservers() {
        notificationToken1 = NotificationCenter.default.observe(name: UIApplication.didEnterBackgroundNotification, object: nil, queue: nil) { [weak self] (notification) in
            self?.sharedResources.synchronizeSharedDefaults()
        }
        notificationToken2 = NotificationCenter.default.observe(name: UIApplication.willTerminateNotification, object: nil, queue: nil) { [weak self] (notification) in
            self?.sharedResources.synchronizeSharedDefaults()
        }
    }
    
    private func isHostInInvertedWhitelist(host: String) -> Bool {
        let invertedDomainsObj = sharedResources.invertedWhitelistContentBlockingObject
        guard let rules = invertedDomainsObj?.rules else { return false }
        
        for rule in rules{
            if rule.ruleText.caseInsensitiveCompare(host) == ComparisonResult.orderedSame && rule.isEnabled.boolValue {
                return true
            }
        }
        
        return false
    }
    
    private func domainObjectIfExistsFromContentBlockingWhitelistFor(host: String) -> AEWhitelistDomainObject? {
        DDLogDebug("(ActionViewController) domainObjectIfExistsFromContentBlockingWhitelistFor:\(host)")
        guard let rules = sharedResources.whitelistContentBlockingRules else { return nil }
        return domainObjectIfExists(host: host, rules: rules)
    }
    
    private func domainObjectIfExists(host: String, rules: NSArray) -> AEWhitelistDomainObject? {
        let predicate = NSPredicate(format: "ruleText CONTAINS[cd] %@", host)
        let filteredRules = rules.filtered(using: predicate)
        
        if filteredRules.count > 0 {
            var obj: AEWhitelistDomainObject? = AEWhitelistDomainObject()
            for rule in filteredRules {
                guard let safeRule = rule as? ASDFilterRule else {
                    obj = nil
                    continue
                }
                obj = AEWhitelistDomainObject.init(rule: safeRule)
                if let _ = AEWhitelistDomainObject.init(rule: safeRule) {
                    break
                }
            }
            return obj
        }
        return nil
    }
    
    private func prepareDataModel() -> NSError? {
        // Init Logger
        ACLLogger.singleton()?.initLogger(sharedResources.sharedAppLogsURL())
        
        let isDebugLogs = sharedResources.isDebugLogs
        DDLogInfo("Start action extension with log level: \(isDebugLogs ? "DEBUG" : "Normal")")
        ACLLogger.singleton()?.logLevel = isDebugLogs ? ACLLDebugLevel : ACLLDefaultLevel
        
        #if DEBUG
        ACLLogger.singleton()?.logLevel = ACLLDebugLevel
        #endif
        
        // Registering standart Defaults
        let appPath = Bundle.main.bundlePath as NSString
        let fullPath = appPath.appendingPathComponent("../../") as String
        guard let path = Bundle(path: fullPath)?.path(forResource: "defaults", ofType: "plist") else {
            DDLogInfo("(ActionViewController) wrong appPath");
            return nil
        }
        if let defs = NSDictionary(contentsOfFile: path) as? [String : Any] {
            DDLogInfo("(ActionViewController) default.plist loaded!")
            sharedResources.sharedDefaults().register(defaults: defs)
        } else {
            DDLogError("(ActionViewController) default.plist was not loaded.")
            return NSError(domain: AEActionErrorDomain, code: AEActionErrorNoDefaults, userInfo: nil)
        }
        //-------------------------------
        
        // Init database
        let dbUrl = sharedResources.sharedResuorcesURL().appendingPathComponent(aeProductionDb)
        asDataBase.initDb(with: dbUrl, upgradeDefaultDb: false, buildVersion: productInfo.buildVersion())
        
        if asDataBase.error != nil {
            DDLogError("(ActionViewController) production DB was not created before.")
            let messageFormat = ACLocalizedString("action_extension_no_configuration_message_format", nil)
            let formattedString = String(format: messageFormat, Constants.aeProductName(), Constants.aeProductName())
            let userInfo = [NSLocalizedDescriptionKey : formattedString]
            return NSError.init(domain: AEActionErrorDomain, code: AEActionErrorNoDb, userInfo: userInfo)
        }
        
        DispatchQueue.main.async {[weak self] in
            guard let sSelf = self else { return }
            //------------ Checking DB status -----------------------------
            if sSelf.asDataBase.error != nil {
                DDLogError("(ActionViewController) production DB was not created before.")
            } else if !sSelf.asDataBase.ready {
                sSelf.dbObserver = sSelf.asDataBase.observe(\.ready, options: .new, changeHandler: { (db, change) in
                    if sSelf.asDataBase.ready{
                        sSelf.antibannerController.start()
                    }
                })
            }
            //--------------------- Start Services ---------------------------
            else {
                sSelf.antibannerController.start()
            }
        }
        return nil
    }
    
    private func updateTheme() {
        view.backgroundColor = theme?.backgroundColor
        theme?.setupNavigationBar(navigationController?.navigationBar)
    }
}
